lib/model/model.go

```diff
@@ -2253,17 +2215,31 @@ func (m *model) AddConnection(conn protocol.Connection, hello protocol.Hello) {
 		// actual close without holding pmut as the connection will call
 		// back into Closed() for the cleanup.
 		closed := m.closed[deviceID]
+		m.fmut.RUnlock()
 		m.pmut.Unlock()
 		oldConn.Close(errReplacingConnection)
 		<-closed
+		// Again, lock fmut before pmut.
+		m.fmut.RLock()
 		m.pmut.Lock()
 	}
 
 	m.conn[deviceID] = conn
 	closed := make(chan struct{})
 	m.closed[deviceID] = closed
 	m.deviceDownloads[deviceID] = newDeviceDownloadState()
-	m.indexSenders[deviceID] = newIndexSenderRegistry(conn, closed, m.Supervisor, m.evLogger)
+	folderStates := make(map[string]*indexHandlerFolderState, len(m.folderCfgs))
+	for id, fcfg := range m.folderCfgs {
+		if fcfg.SharedWith(deviceID) {
+			folderStates[id] = &indexHandlerFolderState{
+				cfg:    fcfg,
+				fset:   m.folderFiles[id],
+				runner: m.folderRunners[id],
+			}
+		}
+	}
+	m.fmut.RUnlock()
+	m.indexHandlers[deviceID] = newIndexHandlerRegistry(conn, m.deviceDownloads[deviceID], closed, folderStates, m.Supervisor, m.evLogger)
```

Comment:
I'm not super excited about exposing the `indexHandlerFolderState` as a thing outside the `indexHandler`, it feels very much like an internal type. The other usages are easy as it's just used to pass essentially two parameters that could be broken out, but this map thing is something else. Perhaps we could solve this by creating the index handler registry and then in the loop calling something like `registry.RegisterFolderState(cfg, fset, runner)`